### PR TITLE
LibGit2 `credentials` callback overhaul

### PR DESCRIPTION
**Pull request by wildart**
Update of the LibGit2 `credentials` callback:
- handle https & ssh connections
- new credential types
- caching support
- `getpass` version for Windows [#8228]

I added a credentials caching to `Pkg.update` which should be helpful when ssh support will arrive,  #16041.

### CODE REVIEW

File: base/libgit2/callbacks.jl [hunk 1]
```diff
@@ -31,74 +31,129 @@ If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayloa
...
-        if isempty(userpass)
-            userpass = prompt("Password for '$username@$url'", password=true)
+    end
+    isusedcreds = isused(creds)
+
+    # use ssh key or ssh-agent
+    if isset(allowed_types, Cuint(Consts.CREDTYPE_SSH_KEY))
+        credid = "ssh://$host"
+        # first try ssh-agent if credentials support its usage
+        if (isdefined(creds, :use_ssh_agent) ? getfield(creds, :use_ssh_agent) : 0xFF) == 0x00
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: no need for this to be ternary

File: base/libgit2/callbacks.jl [hunk 2]
```diff
@@ -31,74 +31,129 @@ If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayloa
...
+            err = ccall((:git_cred_ssh_key_from_agent, :libgit2), Cint,
+                         (Ptr{Ptr{Void}}, Cstring), cred, username_ptr)
+            setfield!(creds, :use_ssh_agent, 0x01) # used ssh-agent ones
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: ones what?

File: base/libgit2/callbacks.jl [hunk 3]
```diff
@@ -31,74 +31,129 @@ If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayloa
...
+                println("ERROR: $errmsg")
+            end
+        end
 
         # if username is not provided, then prompt for it
-        username = if username_ptr == Cstring_NULL
-            prompt("Username for '$url'")
+        username = if username_ptr == Cstring(C_NULL)
+            uname = creds[:user, credid] # check if credentials were already used
+            uname !== nothing && !isusedcreds ? uname : prompt("Username for `$schema$host'")
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: why the backtick?

File: base/libgit2/types.jl [hunk 4]
```diff
@@ -52,10 +52,29 @@ end
...
-password(p::AbstractPayload) = throw(AssertionError("Function 'password' is not implemented for type $(typeof(p))"))
-isused(p::AbstractPayload) = throw(AssertionError("Function 'isused' is not implemented for type $(typeof(p))"))
-setused!(p::AbstractPayload,v::Bool) = throw(AssertionError("Function 'setused!' is not implemented for type $(typeof(p))"))
+
+"Abstract credentials payload"
+abstract AbstractCredentials <: AbstractPayload
+"Returns a credentials parameter"
+function Base.getindex(p::AbstractCredentials, keys...)
+    for k in keys
+        ks = symbol(k)
```
Comment (tkelman) [inline base/libgit2/types.jl]: deprecated

File: base/libgit2/types.jl [hunk 5]
```diff
@@ -612,13 +631,54 @@ function getobjecttype(obj_type::Cint)
...
     pass::AbstractString
-    used::Bool
-    UserPasswordCredentials(u::AbstractString,p::AbstractString) = new(u,p,false)
+    count::Int
+    use_ssh_agent::UInt8
+    UserPasswordCredentials(u::AbstractString,p::AbstractString) = new(u,p,3,zero(UInt8))
+end
+function isused(p::UserPasswordCredentials)
+    p.count <= 0 && return true
+    p.count -= 1
```
Comment (tkelman) [inline base/libgit2/types.jl]: what is this counting, and why does `isused` decrement it? it should be called `isused` if it's mutating

File: base/libgit2/utils.jl [hunk 6]
```diff
@@ -10,14 +12,54 @@ function version()
...
+        end
+        return bytestring(pointer(p), plen)
+    finally
+        fill!(p, 0) # don't leave password in memory
+    end
+
+    return ""
+end
+
+@unix_only getpass(prompt::AbstractString) = bytestring(pointer_to_string(ccall(:getpass, Cstring, (Cstring,), prompt), true))
```
Comment (tkelman) [inline base/libgit2/utils.jl]: wrap long line

File: base/libgit2/utils.jl [hunk 7]
```diff
@@ -15,10 +15,39 @@ isset(val::Integer, flag::Integer) = (val & flag == flag)
 reset(val::Integer, flag::Integer) = (val &= ~flag)
 toggle(val::Integer, flag::Integer) = (val |= flag)
 
+@windows_only function getpass(prompt::AbstractString)
```
Comment (tkelman) [inline base/libgit2/utils.jl]: this isn't libgit2 specific

File: base/libgit2/types.jl (hunk 5 above)
Comment (wildart) [inline base/libgit2/types.jl]: This is because of the specifics of libgit2 authentication procedure: when authentication is failed, libgit2 calls credentials callback again, but there is no information whether authentication was successful or not. So, in order not to stuck in infinite loop by using same faulty credentials simple counting strategy is implemented.

Comment (tkelman) [inline base/libgit2/types.jl]: **please** document things like that with copious comments.

**Comment by wildart** [comment]
I added description of  libgit2 authentication procedure to `LibGit2.credentials_callback` documentation and some `Credentials` types.

File: base/libgit2/callbacks.jl [hunk 8]
```diff
@@ -31,74 +31,131 @@ If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayloa
 
 For `LibGit2.Consts.CREDTYPE_USERPASS_PLAINTEXT` type, if payload contains fields: `user` & `pass` they are used to create authentication credentials.
 In addition, if payload has field `used` it can be set to `true` to indicate that payload was used and abort callback with error. This behavior is required to avoid repeated authentication calls with incorrect credentials.
+
+Order of credential checks (if supported):
+- ssh key pair (ssh-agent if specified in `payload`)
+- plain text
+
+**Note**: Due to the specifics of `libgit2` authentication procedure, when authentication is failed, this functions is called again without any indication whether authentication was successful or not. In order not to stuck in infinite loop by repeatedly using same faulty credentials, use a call counting strategy to counteract this behaviour.
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: wrap docstrings. "when authentication fails"

File: base/libgit2/callbacks.jl [hunk 9]
```diff
@@ -31,74 +31,131 @@ If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayloa
...
-        if isempty(userpass)
-            userpass = prompt("Password for '$username@$url'", password=true)
+    end
+    isusedcreds = isused(creds)
+
+    # use ssh key or ssh-agent
+    if isset(allowed_types, Cuint(Consts.CREDTYPE_SSH_KEY))
+        credid = "ssh://$host"
+        # first try ssh-agent if credentials support its usage
+        if isdefined(creds, :use_ssh_agent) && getfield(creds, :use_ssh_agent) == 0x00
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: can't the `getfield` and `setfield!` just using `creds.use_ssh_agent` if it's guarded by an `isdefined`?

File: base/libgit2/callbacks.jl [hunk 10]
```diff
@@ -31,74 +31,131 @@ If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayloa
...
         else
-            keydef = homedir()*"/.ssh/id_rsa.pub"
-            prompt("Public key location", default=keydef)
+            keydef = creds[:pubkey, credid] # check if credentials were already used
+            if keydef !== nothing && !isusedcreds
+                keydef # use cached value
+            else
+                keydef = homedir()*"/.ssh/id_rsa.pub"
+                prompt("Public key location for '$schema$username@$host'", default=keydef)
+            end
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: this "`hasenv` otherwise `prompt`" pattern looks like it should be refactored into a common function

File: base/libgit2/utils.jl [hunk 11]
```diff
@@ -10,14 +12,55 @@ function version()
 end
 
 isset(val::Integer, flag::Integer) = (val & flag == flag)
+reset(val::Integer, flag::Integer) = (val &= ~flag)
+toggle(val::Integer, flag::Integer) = (val |= flag)
+
+@windows_only function getpass(prompt::AbstractString)
```
Comment (tkelman) [inline base/libgit2/utils.jl]: this isn't libgit2-specific meaning it should be in a different module

File: base/libgit2/types.jl [hunk 12]
```diff
@@ -612,13 +631,60 @@ function getobjecttype(obj_type::Cint)
...
-isused(p::UserPasswordCredentials) = p.used
-setused!(p::UserPasswordCredentials, val::Bool) = (p.used = val)
+    count::Int            # authentication failure protection count
+    use_ssh_agent::UInt8  # used for ssh-agent authentication
+    UserPasswordCredentials(u::AbstractString,p::AbstractString) = new(u,p,3,zero(UInt8))
+end
+"Checks if credentials were used or failed authentication, see `LibGit2.credentials_callback`"
+function isused(p::UserPasswordCredentials)
+    p.count <= 0 && return true
+    p.count -= 1
```
Comment (tkelman) [inline base/libgit2/types.jl]: given this is mutating, it should be either `isused!`, or probably better, `checkused!`

File: base/libgit2/callbacks.jl (hunk 10 above)
Comment (wildart) [inline base/libgit2/callbacks.jl]: it is going to be messy - too many parameters to pass

File: base/libgit2/callbacks.jl [hunk 13]
```diff
@@ -29,14 +29,19 @@ end
...
+In addition, if payload has field `used` it can be set to `true` to indicate that payload was used
+and abort callback with error. This behavior is required to avoid repeated authentication calls with incorrect credentials.
 
 Order of credential checks (if supported):
 - ssh key pair (ssh-agent if specified in `payload`)
 - plain text
 
-**Note**: Due to the specifics of `libgit2` authentication procedure, when authentication is failed, this functions is called again without any indication whether authentication was successful or not. In order not to stuck in infinite loop by repeatedly using same faulty credentials, use a call counting strategy to counteract this behaviour.
+**Note**: Due to the specifics of `libgit2` authentication procedure, when authentication fails,
+this functions is called again without any indication whether authentication was successful or not.
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: this function

File: base/libgit2/callbacks.jl [hunk 14]
```diff
@@ -29,14 +29,19 @@ end
...
+In order not to stuck in infinite loop by repeatedly using same faulty credentials,
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: To avoid an infinite loop from repeatedly...

File: base/libgit2/callbacks.jl (hunk 10 above)
Comment (tkelman) [inline base/libgit2/callbacks.jl]: only 2 or 3. repeating the exact same logic with different variables is messier.

File: base/libgit2/types.jl [hunk 15]
```diff
@@ -52,10 +52,29 @@ end
...
+    end
+    return nothing
+end
+"Sets credentials with `key` parameter with value"
+function Base.setindex!(p::AbstractCredentials, val, keys...)
+    for k in keys
+        ks = Symbol(k)
+        isdefined(p, ks) && setfield!(p, ks, val)
+    end
+    return nothing
```
Comment (tkelman) [inline base/libgit2/types.jl]: I think `setindex!` should return the modified object back, by convention

File: base/libgit2/types.jl [hunk 16]
```diff
@@ -612,13 +631,60 @@ function getobjecttype(obj_type::Cint)
...
+    Dict(:user=>1, :pass=>2, :pubkey=>3, :prvkey=>4, :usesshagent=>5)
+
+"Credentials that support caching"
+type CachedCredentials <: AbstractCredentials
+    cred::Dict{AbstractString,Vector{AbstractString}}
+    count::Int            # authentication failure protection count
+    CachedCredentials() = new(Dict{AbstractString,Vector{AbstractString}}(),3)
+end
+"Returns specific credential parameter value: first index is a credential parameter name, second index is a host name (with schema)"
+function Base.getindex(p::CachedCredentials, keys...)
```
Comment (tkelman) [inline base/libgit2/types.jl]: why is this varargs?

**Comment by tkelman** [comment]
Something in the tests here is repeatably freezing on AppVeyor.

**Comment by wildart** [comment]
It is `libgit2-online` test that suddenly became interactive and stick in prompt.

File: test/libgit2-online.jl [hunk 17]
```diff
@@ -21,12 +21,17 @@ mktempdir() do dir
                 finalize(repo)
             end
         #end
+        try
         #@testset "with incorrect url" begin
             repo_path = joinpath(dir, "Example2")
             # credential are required because github try authenticate on uknown repo
```
Comment (tkelman) [inline test/libgit2-online.jl]: unknown

File: test/libgit2-online.jl [hunk 18]
```diff
@@ -21,12 +21,17 @@ mktempdir() do dir
...
-            x = Nullable(LibGit2.UserPasswordCredentials("X","X"))
-            @test_throws LibGit2.Error.GitError LibGit2.clone(https_prefix*repo_url*randstring(10), repo_path, payload=x)
+            cred = LibGit2.UserPasswordCredentials("","") # empty credentials cause authentication error
+            LibGit2.clone(https_prefix*repo_url*randstring(10), repo_path, payload=Nullable(cred))
```
Comment (tkelman) [inline test/libgit2-online.jl]: is expected to always fail? if so it should throw a different type of exception at the end of the try so it doesn't silently succeed

File: test/libgit2-online.jl (hunk 17 above)
Comment (wildart) [inline test/libgit2-online.jl]: my keybord is dying :skull:

File: test/libgit2-online.jl (hunk 18 above)
Comment (wildart) [inline test/libgit2-online.jl]: yes, you're right

**Comment by wildart** [comment]
I realized that public key isn't necessary in SSH authorization. OpenSSL (as well as MbedTLS) can extract public key from private, see [here](https://www.libssh2.org/libssh2_userauth_publickey_fromfile_ex.html). So I'll remove public key prompt.

**Comment by wildart** [comment]
Above proposition went south. Apparently, system's `libssh2` build against `libgcrypt` which does not support extraction of a public key from a private one. I guess it's another reason to move to mbedtls crypto backend for `libssh2`.

**Comment by tkelman** [comment]
Any update from libssh2 on reviewing your PR? I tried an early version of it which didn't work, haven't tested your latest commits yet though.

**Comment by wildart** [comment]
Judging from accepted libssh2 PRs, this could take some time.

**Comment by tkelman** [comment]
If it works we can use a locally patched version for a little while, but I don't want to be indefinitely shipping unreviewed crypto code.

**Comment by wildart** [comment]
Ready to merge.

File: base/libgit2/callbacks.jl [hunk 19]
```diff
@@ -29,76 +29,154 @@ end
 Function provides different credential acquisition functionality w.r.t. a connection protocol.
 If payload is provided then `payload_ptr` should contain `LibGit2.AbstractPayload` object.
 
-For `LibGit2.Consts.CREDTYPE_USERPASS_PLAINTEXT` type, if payload contains fields: `user` & `pass` they are used to create authentication credentials.
-In addition, if payload has field `used` it can be set to `true` to indicate that payload was used and abort callback with error. This behavior is required to avoid repeated authentication calls with incorrect credentials.
+For `LibGit2.Consts.CREDTYPE_USERPASS_PLAINTEXT` type, if payload contains fields:
+`user` & `pass` they are used to create authentication credentials.
+In addition, if payload has field `used` it can be set to `true` to indicate that payload was used
+and abort callback with error. This behavior is required to avoid repeated authentication calls with incorrect credentials.
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: What do you mean by "to indicate that payload was used and abort callback with error"? If the field is set to true, where does the callback get aborted?

File: base/libgit2/callbacks.jl [hunk 20]
```diff
@@ -29,76 +29,154 @@ end
...
+
+Order of credential checks (if supported):
+- ssh key pair (ssh-agent if specified in `payload`)
+- plain text
+
+**Note**: Due to the specifics of `libgit2` authentication procedure, when authentication fails,
+this functions is called again without any indication whether authentication was successful or not.
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: this function

File: base/libgit2/callbacks.jl [hunk 21]
```diff
@@ -29,76 +29,154 @@ end
...
-                userpass = password(payload)
-                setused!(payload, true)
-            end
+    # parse url for schema and host
+    urlparts = match(urlmatcher, url)
+    schema = urlparts.captures[1]
+    host = urlparts.captures[5]
+    schema = schema === nothing ? "" : schema*"://"
+
+    # get credentials object form payload ref
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: from? or get object then form?

File: base/libgit2/callbacks.jl [hunk 22]
```diff
@@ -29,76 +29,154 @@ end
...
-        if isempty(userpass)
-            userpass = prompt("Password for '$username@$url'", password=true)
+    end
+    isusedcreds = checkused!(creds)
+
+    # use ssh key or ssh-agent
+    if isset(allowed_types, Cuint(Consts.CREDTYPE_SSH_KEY))
+        credid = "ssh://$host"
+
+        # set ssh-agent triger for first use
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: trigger

File: base/libgit2/types.jl [hunk 23]
```diff
@@ -52,10 +52,29 @@ end
...
+
+"Abstract credentials payload"
+abstract AbstractCredentials <: AbstractPayload
+"Returns a credentials parameter"
+function Base.getindex(p::AbstractCredentials, keys...)
+    for k in keys
+        ks = Symbol(k)
+        isdefined(p, ks) && return getfield(p, ks)
+    end
+    return nothing
```
Comment (tkelman) [inline base/libgit2/types.jl]: Should this be a KeyError or UndefRefError? If this is an associative structure, should it implement `haskey` ?

File: test/libgit2-online.jl [hunk 24]
```diff
@@ -21,12 +21,18 @@ mktempdir() do dir
                 finalize(repo)
             end
         #end
+        try
         #@testset "with incorrect url" begin
             repo_path = joinpath(dir, "Example2")
-            # credential are required because github try authenticate on uknown repo
-            x = Nullable(LibGit2.UserPasswordCredentials("X","X"))
-            @test_throws LibGit2.Error.GitError LibGit2.clone(https_prefix*repo_url*randstring(10), repo_path, payload=x)
+            # credential are required because github try authenticate on unknown repo
```
Comment (tkelman) [inline test/libgit2-online.jl]: credentials

File: test/libgit2-online.jl [hunk 25]
```diff
@@ -21,12 +21,18 @@ mktempdir() do dir
...
+            cred = LibGit2.UserPasswordCredentials("","") # empty credentials cause authentication error
+            LibGit2.clone(https_prefix*repo_url*randstring(10), repo_path, payload=Nullable(cred))
+            error("unexpected")
         #end
```
Comment (tkelman) [inline test/libgit2-online.jl]: the commented end-of-testset should presumably be after the catch, not in the middle of it

File: base/libgit2/callbacks.jl (hunk 19 above)
Comment (wildart) [inline base/libgit2/callbacks.jl]: this part of docs is irrelevant, I'll rewrite it.

File: base/libgit2/types.jl (hunk 23 above)
Comment (wildart) [inline base/libgit2/types.jl]: I did not want to impose any particular structure on `AbstractCredentials` subtypes.

File: base/libgit2/callbacks.jl [hunk 26]
```diff
@@ -27,21 +27,27 @@ end
...
+- ssh key pair (ssh-agent if specified in payload's `usesshagent` field)
 - plain text
 
 **Note**: Due to the specifics of `libgit2` authentication procedure, when authentication fails,
-this functions is called again without any indication whether authentication was successful or not.
+this function is called again without any indication whether authentication was successful or not.
 To avoid an infinite loop from repeatedly using the same faulty credentials,
-use a call counting strategy to counteract this behaviour.
+`checkused!` function can be called to test if credentials were used if call returns `true` value.
+Used credentials trigger user prompt for (r)entering required information.
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: (re)entering

File: base/libgit2/callbacks.jl [hunk 27]
```diff
@@ -27,21 +27,27 @@ end
...
+`UserPasswordCredentials` and `CachedCredentials` are implemented a call counting strategy
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: "implement" or "are implemented using"

File: base/libgit2/callbacks.jl [hunk 28]
```diff
@@ -27,21 +27,27 @@ end
...
+that counteracts faulty credentials repeated usage.
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: that prevents repeated usage of faulty credentials

File: test/libgit2-online.jl [hunk 29]
```diff
@@ -21,28 +21,29 @@ mktempdir() do dir
...
+
         #@testset "with incorrect url" begin
-            repo_path = joinpath(dir, "Example2")
-            # credential are required because github try authenticate on unknown repo
-            cred = LibGit2.UserPasswordCredentials("","") # empty credentials cause authentication error
-            LibGit2.clone(https_prefix*repo_url*randstring(10), repo_path, payload=Nullable(cred))
-            error("unexpected")
+            try
+                repo_path = joinpath(dir, "Example2")
+                # credentials are required because github tries authenticate on unknown repo
```
Comment (tkelman) [inline test/libgit2-online.jl]: tries to authenticate

File: test/libgit2-online.jl [hunk 30]
```diff
@@ -21,28 +21,29 @@ mktempdir() do dir
...
-            @test isa(ex, LibGit2.Error.GitError)
-            @test ex.code == LibGit2.Error.EAUTH
-        end
 
-        try
-            #@testset "with 'ssh' protocol (by default is not supported)" begin
+        #@testset "with 'ssh' protocol (by default is not supported)" begin
+            try
                 repo_path = joinpath(dir, "Example3")
                 @test_throws LibGit2.Error.GitError LibGit2.clone(ssh_prefix*repo_url, repo_path)
```
Comment (tkelman) [inline test/libgit2-online.jl]: is this expected to always fail, or will it depend on whether libgit2 is built with ssh support? we can check that now with features, right?

Comment (wildart) [inline test/libgit2-online.jl]: We deal with this test when ssh will be available. Yes, features can be checked to see which protocols are supported, but that is not part of this PR. The rework of the url handling, similar to #13879, should include it.

Comment (tkelman) [inline test/libgit2-online.jl]: can you add a todo comment?

File: test/libgit2-online.jl [hunk 31]
```diff
@@ -21,22 +21,29 @@ mktempdir() do dir
...
         #end
+
         #@testset "with incorrect url" begin
-            repo_path = joinpath(dir, "Example2")
-            # credential are required because github try authenticate on uknown repo
-            x = Nullable(LibGit2.UserPasswordCredentials("X","X"))
-            @test_throws LibGit2.Error.GitError LibGit2.clone(https_prefix*repo_url*randstring(10), repo_path, payload=x)
+            try
+                repo_path = joinpath(dir, "Example2")
+                # credentials are required because github tries authenticate on unknown repo
```
Comment (tkelman) [inline test/libgit2-online.jl]: tries to authenticate

File: base/libgit2/callbacks.jl [hunk 32]
```diff
@@ -27,78 +27,162 @@ end
...
+
+Order of credential checks (if supported):
+- ssh key pair (ssh-agent if specified in payload's `usesshagent` field)
+- plain text
+
+**Note**: Due to the specifics of `libgit2` authentication procedure, when authentication fails,
+this function is called again without any indication whether authentication was successful or not.
+To avoid an infinite loop from repeatedly using the same faulty credentials,
+`checkused!` function can be called to test if credentials were used if call returns `true` value.
+Used credentials trigger user prompt for (r)entering required information.
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: (re)entering

File: base/libgit2/callbacks.jl [hunk 33]
```diff
@@ -27,78 +27,162 @@ end
...
+`UserPasswordCredentials` and `CachedCredentials` are implemented a call counting strategy
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: are implemented using

File: base/libgit2/callbacks.jl [hunk 34]
```diff
@@ -27,78 +27,162 @@ end
...
+that counteracts faulty credentials repeated usage.
```
Comment (tkelman) [inline base/libgit2/callbacks.jl]: that prevents repeated usage of faulty credentials

(I'm repeating myself, github hid these but they weren't addressed)

**Comment by tkelman** [comment]
lgtm after squashing a bit so that none of the intermediate commits would have the hanging issue

**Comment by tkelman** [comment]
Given #16373 should we hold back on merging this to allow a little bit more debugging?

**Comment by wildart** [comment]
Yep, I need to look at this.

**Comment by wildart** [comment]
It's ready to merge.

**Comment by stevengj** [comment]
Shouldn't `getpass` be exported and documented?
